Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift

```diff
@@ -198,8 +207,16 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
             logger.debug("Connection pool exhausted on this event loop, adding request to waitlist")
             let promise = eventLoop.makePromise(of: Source.Connection.self)
             self.waiters.append((logger, promise))
-            // return waiter
-            return promise.futureResult
+            
+            let task = eventLoop.scheduleTask(in: self.requestTimeout) { [weak self, logger, promise] in
+                logger.error("Connection request timed out. This might indicate a connection deadlock in your application.")
+                if let idx = self?.waiters.firstIndex(where: { let p = $0.1; return p.futureResult === promise.futureResult }) {
+                    self?.waiters.remove(at: idx)
+                }
+                promise.fail(ConnectionPoolTimeoutError.connectionRequestTimeout)
+            }
+            
+            return promise.futureResult.always { _ in task.cancel() }
```

Comment:
Is cancelling a task that has already completed fine?

Reply:
I did a quick test where the task would always finish before cancelling and didn't see anything going wrong, so guessing it's fine. Not 100% sure though.

cc @Lukasa